File: product-microservice/api/solid_state_drive_api.go
package api

import (
	"product/dto"
	"product/dto/filter"
	"product/mapper"
	"product/service"
	"strconv"

	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)


type SolidStateDriveAPI struct {
	ISolidStateDriveService service.ISolidStateDriveService
}

func NewSolidStateDriveAPI(solidStateDriveService service.ISolidStateDriveService) SolidStateDriveAPI {
	return SolidStateDriveAPI{ISolidStateDriveService: solidStateDriveService}
}

func (solidStateDriveApi *SolidStateDriveAPI) GetAll(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }
	
	solidStateDrives := solidStateDriveApi.ISolidStateDriveService.GetAll(page, pageSize)
	c.JSON(http.StatusOK, mapper.ToSolidStateDriveDTOs(solidStateDrives))
}

func (solidStateDriveApi *SolidStateDriveAPI) GetByID(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	solidStateDrive, err := solidStateDriveApi.ISolidStateDriveService.GetById(id)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToSolidStateDriveDTO(solidStateDrive))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (solidStateDriveApi *SolidStateDriveAPI) SearchByName(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }
	
	solidStateDrives, err := solidStateDriveApi.ISolidStateDriveService.SearchByName(page, pageSize, c.Query("name"))
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToSolidStateDriveDTOs(solidStateDrives))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (solidStateDriveApi *SolidStateDriveAPI) Filter(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }

	var filter filter.SolidStateDriveFilter
	err = c.BindJSON(&filter)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	solidStateDrives, err := solidStateDriveApi.ISolidStateDriveService.Filter(page, pageSize, filter)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToSolidStateDriveDTOs(solidStateDrives))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (solidStateDriveApi *SolidStateDriveAPI) GetCapacities(c *gin.Context) {
	capacities := solidStateDriveApi.ISolidStateDriveService.GetCapacities()
	c.JSON(http.StatusOK, capacities)
}

func (solidStateDriveApi *SolidStateDriveAPI) GetForms(c *gin.Context) {
	forms := solidStateDriveApi.ISolidStateDriveService.GetForms()
	c.JSON(http.StatusOK, forms)
}

func (solidStateDriveApi *SolidStateDriveAPI) GetManufacturers(c *gin.Context) {
	manufacturers := solidStateDriveApi.ISolidStateDriveService.GetManufacturers()
	c.JSON(http.StatusOK, manufacturers)
}

func (solidStateDriveApi *SolidStateDriveAPI) GetMaxSequentialReads(c *gin.Context) {
	maxSequentialReads := solidStateDriveApi.ISolidStateDriveService.GetMaxSequentialReads()
	c.JSON(http.StatusOK, maxSequentialReads)
}

func (solidStateDriveApi *SolidStateDriveAPI) GetMaxSequentialWrites(c *gin.Context) {
	maxSequentialWrites := solidStateDriveApi.ISolidStateDriveService.GetMaxSequentialWrites()
	c.JSON(http.StatusOK, maxSequentialWrites)
}

func (solidStateDriveApi *SolidStateDriveAPI) Create(c *gin.Context) {
	var solidStateDriveDTO dto.SolidStateDriveDTO
	err := c.BindJSON(&solidStateDriveDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	solidStateDrive := mapper.ToSolidStateDrive(solidStateDriveDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	error := solidStateDriveApi.ISolidStateDriveService.Create(solidStateDrive)

	if error == nil {
		c.JSON(http.StatusOK, "Solid state drive stored successfully.")
	} else {
		c.JSON(http.StatusBadRequest, error.Error())
	}
}

func (solidStateDriveApi *SolidStateDriveAPI) Update(c *gin.Context) {
	var solidStateDriveDTO dto.SolidStateDriveDTO
	err := c.BindJSON(&solidStateDriveDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	error := solidStateDriveApi.ISolidStateDriveService.Update(solidStateDriveDTO)

	if error == nil {
		c.JSON(http.StatusOK, "Solid state drive updated successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, error.Error())
	} 
}

func (solidStateDriveApi *SolidStateDriveAPI) Delete(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	error := solidStateDriveApi.ISolidStateDriveService.Delete(id)

	if error == nil {
		c.JSON(http.StatusOK, "Solid state drive deleted successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, error.Error())
	}
}

File: api-gateway/api/product_microservice/solid_state_drive_api.py
import requests
from flask import request, jsonify
from utils import token_utils, role
from utils.routes.product_microservice import solid_state_drive_api_routes


def get_all():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    try:
        r = requests.get(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                         solid_state_drive_api_routes.GET_ALL +
                         "?page={}&pageSize={}".format(page, page_size))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_number_of_records():
    try:
        r = requests.get(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                         solid_state_drive_api_routes.GET_NUMBER_OF_RECORDS)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_by_id(id):
    try:
        r = requests.get(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                         solid_state_drive_api_routes.GET_BY_ID + "/{}".format(id))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def search_by_name():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    name = args.get("name")
    try:
        r = requests.get(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                         solid_state_drive_api_routes.SEARCH_BY_NAME +
                         "?page={}&pageSize={}&name={}".format(page, page_size, name))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_number_of_records_search():
    args = request.args.to_dict()
    name = args.get("name")
    try:
        r = requests.get(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                         solid_state_drive_api_routes.GET_NUMBER_OF_RECORDS_SEARCH + "?name={}".format(name))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def filter():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    data = request.json
    try:
        r = requests.post(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                          solid_state_drive_api_routes.FILTER +
                          "?page={}&pageSize={}".format(page, page_size), json=data)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_number_of_records_filter():
    data = request.json
    try:
        r = requests.post(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                          solid_state_drive_api_routes.GET_NUMBER_OF_RECORDS_FILTER, json=data)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_manufacturers():
    try:
        r = requests.get(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                         solid_state_drive_api_routes.GET_MANUFACTURERS)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_forms():
    try:
        r = requests.get(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                         solid_state_drive_api_routes.GET_FORMS)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_capacities():
    try:
        r = requests.get(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                         solid_state_drive_api_routes.GET_CAPACITIES)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_max_sequential_reads():
    try:
        r = requests.get(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                         solid_state_drive_api_routes.GET_MAX_SEQUENTIAL_READS)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_max_sequential_writes():
    try:
        r = requests.get(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                         solid_state_drive_api_routes.GET_MAX_SEQUENTIAL_WRITES)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def create():
    data = request.json
    headers = request.headers
    try:
        r = requests.post(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                          solid_state_drive_api_routes.CREATE, json=data, headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def update():
    data = request.json
    headers = request.headers
    try:
        r = requests.put(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                         solid_state_drive_api_routes.UPDATE, json=data, headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def delete(id):
    headers = request.headers
    try:
        r = requests.delete(solid_state_drive_api_routes.BASE + solid_state_drive_api_routes.API +
                            solid_state_drive_api_routes.DELETE + "/{}".format(id), headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp
